### PR TITLE
PHOENIX-5539 Full row index write at the last write phase for mutable…

### PR DESCRIPTION
**Pull request by kadirozde**
… global indexes

### CODE REVIEW

File: phoenix-core/src/it/java/org/apache/phoenix/end2end/index/GlobalIndexCheckerIT.java [hunk 1]
```diff
@@ -257,6 +266,44 @@ public void testFailPostIndexRowUpdate() throws Exception {
...
+            conn.commit();
+            IndexRegionObserver.setFailPostIndexUpdatesForTesting(false);
+            Table hIndexTable = conn.unwrap(PhoenixConnection.class).getQueryServices()
+                    .getTable(Bytes.toBytes(indexTableName));
+            long indexCnt = TestUtil.getRowCount(hIndexTable, false);
+            assertEquals(1, indexCnt);
+            assertEquals(true, verifyRowsForEmptyColValue(conn, indexTableName, IndexRegionObserver.UNVERIFIED_BYTES));
+            // check that in the first phase we don't send the full row.
+            // We count the num of cells for this
+            Scan s = new Scan();
```
Comment (gokceni) [inline phoenix-core/src/it/java/org/apache/phoenix/end2end/index/GlobalIndexCheckerIT.java]: nit: nice to have is to have this as a separate function to call. But don't worry about it

File: phoenix-core/src/main/java/org/apache/phoenix/hbase/index/IndexRegionObserver.java [hunk 2]
```diff
@@ -555,34 +570,33 @@ private void prepareIndexMutations(ObserverContext<RegionCoprocessorEnvironment>
...
+                      // going through concurrent updates
+                      ImmutableBytesPtr rowKey = new ImmutableBytesPtr(next.getSecond());
+                      if (!context.pendingRows.contains(rowKey)) {
+                          if (m instanceof Put) {
+                              // Remove the empty column prepared by Index codec as we need to change its value
+                              removeEmptyColumn(m, emptyCF, emptyCQ);
+                              ((Put) m).addColumn(emptyCF, emptyCQ, now, VERIFIED_BYTES);
                               context.intermediatePostIndexUpdates.add(next);
+                          } else {
+                              if (!context.pendingRows.contains(rowKey)) {
```
Comment (gokceni) [inline phoenix-core/src/main/java/org/apache/phoenix/hbase/index/IndexRegionObserver.java]: Do we need this if? There is if in 590 so it should be ok to remove this

Comment (kadirozde) [reply phoenix-core/src/main/java/org/apache/phoenix/hbase/index/IndexRegionObserver.java]: Will do that.

File: phoenix-core/src/it/java/org/apache/phoenix/end2end/index/GlobalIndexCheckerIT.java (hunk 1 above)
Comment (kadirozde) [reply phoenix-core/src/it/java/org/apache/phoenix/end2end/index/GlobalIndexCheckerIT.java]: It is a good idea. I will do that.

File: phoenix-core/src/main/java/org/apache/phoenix/hbase/index/IndexRegionObserver.java [hunk 3]
```diff
@@ -555,33 +570,30 @@ private void prepareIndexMutations(ObserverContext<RegionCoprocessorEnvironment>
                   // add the VERIFIED cell, which is the empty cell
                   Mutation m = next.getFirst().getFirst();
                   boolean rebuild = PhoenixIndexMetaData.isIndexRebuild(m.getAttributesMap());
-                  long ts = getMaxTimestamp(m);
                   if (rebuild) {
                       if (m instanceof Put) {
+                          long ts = getMaxTimestamp(m);
+                          // Remove the empty column prepared by Index codec as we need to change its value
+                          removeEmptyColumn(m, emptyCF, emptyCQ);
```
Comment (BinShi-SecularBird) [inline phoenix-core/src/main/java/org/apache/phoenix/hbase/index/IndexRegionObserver.java]: It seems that we first remove the existing "empty" column from an array -- the cell list of the column family then add a new "empty" column. Is it necessary have an optimization here - replace the old cell with the new cell in place in the array?

File: phoenix-core/src/main/java/org/apache/phoenix/hbase/index/IndexRegionObserver.java [hunk 4]
```diff
@@ -555,33 +570,30 @@ private void prepareIndexMutations(ObserverContext<RegionCoprocessorEnvironment>
...
+                      context.preIndexUpdates.add(new Pair <>(unverifiedPut, next.getFirst().getSecond()));
+                      // Ignore post index updates (i.e., the third write phase updates) for this row if it is
+                      // going through concurrent updates
+                      ImmutableBytesPtr rowKey = new ImmutableBytesPtr(next.getSecond());
+                      if (!context.pendingRows.contains(rowKey)) {
+                          if (m instanceof Put) {
+                              // Remove the empty column prepared by Index codec as we need to change its value
+                              removeEmptyColumn(m, emptyCF, emptyCQ);
+                              ((Put) m).addColumn(emptyCF, emptyCQ, now, VERIFIED_BYTES);
+                              context.intermediatePostIndexUpdates.add(next);
```
Comment (BinShi-SecularBird) [inline phoenix-core/src/main/java/org/apache/phoenix/hbase/index/IndexRegionObserver.java]: Move "context.intermediatePostIndexUpdates.add(next);" outside of if ... else?

Comment (kadirozde) [reply phoenix-core/src/main/java/org/apache/phoenix/hbase/index/IndexRegionObserver.java]: ok. I will do that

File: phoenix-core/src/main/java/org/apache/phoenix/hbase/index/IndexRegionObserver.java (hunk 3 above)
Comment (kadirozde) [reply phoenix-core/src/main/java/org/apache/phoenix/hbase/index/IndexRegionObserver.java]: No need to transfer an unnecessary cell mutation over the network. Also, giving the same cell with two different values is confusing, although HBase takes the latest one.

Comment (BinShi-SecularBird) [reply phoenix-core/src/main/java/org/apache/phoenix/hbase/index/IndexRegionObserver.java]: That isn't what I meant. I meant overriding the old cell with the new cell in the cell list in place and avoiding removing an element in an array.

Comment (kadirozde) [reply phoenix-core/src/main/java/org/apache/phoenix/hbase/index/IndexRegionObserver.java]: There is no HBase API to do that as far as I know, that is why I have to remove it.

**Comment by gjacoby126** [comment]
JIRA already committed, closing.

**Pull request closed without merging**